QNX Compatibility

JaiXJM-BB opened the pull request:
Compatibility for the QNX Operating System, versions 7.1 and 8.0. As discussed in #1914 .

**Changes in Detail:**
Each change will have a GitHub comment attached explaining why it is necessary. Below is an overview of the changes as a whole.
All changes are wrapped in appropriate `if(QNX)` or `if defined(__QNX__)` statements.

- About half of the CMake changes were to adjust compiling options for all builds to accommodate QNX. These generally already existed for other OSes, and simply needed a check for QNX tacked on as well. 
- Some install commands were changed to respect `CMAKE_INSTALL_INCLUDEDIR` instead of using `include/`.
- The rest of the CMake changes enable the compilation of tests. This is so that tests can be run on QNX target, as cross-compilation is required and thus using CTest is unideal.
-- There is a CACHE variable added to support this which only applies if the target is a QNX system. `QNX_TARGET_DATASET_DIR` is set to wherever the examples/Data directory will be cloned to on the target via `-DQNX_TARGET_DATASET:STRING=path/to/dir" when compiling tests to avoid some file location errors. 

- There are two non-CMake changes here.
-- One adds to a gcc bugfix that is slightly worse on qcc (QNX Compiler)
-- The second switches a placeholder in config.h.in to a different filepaths on QNX.

JaiXJM-BB committed:
GitLab version (branch QNX_7.1_v4.1.1)

JaiXJM-BB committed:
ADDED: Build tested with target `install` or `all` if cross compiling for QNX

JaiXJM-BB committed:
ADDED: Test Installation when building with QNX (Review this for upstreaming)

JaiXJM-BB committed:
Updated: Build tests, fixed some unit tests. Floating points still off.

JaiXJM-BB committed:
Tests Fix

JaiXJM-BB committed:
QNX Changes

JaiXJM-BB committed:
QNX for develop branch

* GitLab version (branch QNX_7.1_v4.1.1)

* ADDED: Build tested with target `install` or `all` if cross compiling for QNX

* ADDED: Test Installation when building with QNX (Review this for upstreaming)

* UPDATED: Build tests, fixed some unit tests. Floating points still off.

* UPDATED: Tests Fix

* UPDATED: all non-serialization tests working.

* QNX 8.0: Working version.

* REMOVED: Removal of test prints

* UPDATED: formatting to match, removed commented out testing lines

JaiXJM-BB committed:
erroneous change

JaiXJM-BB committed:
fixing spacing

JaiXJM-BB committed:
fixing spacing issue

JaiXJM-BB committed:
Fixed broken merge

JaiXJM-BB committed:
Removed erroneous text

JaiXJM-BB commented on cmake/GtsamTesting.cmake:
```diff
@@ -197,9 +197,11 @@ macro(gtsamAddTestsGlob_impl groupName globPatterns excludedFiles linkLibraries)
-				set_target_properties(${script_name} PROPERTIES EXCLUDE_FROM_ALL ON)
+				if(NOT QNX OR NOT DEFINED ENV{QNX_BUILD_TESTS})
+					set_target_properties(${script_name} PROPERTIES EXCLUDE_FROM_ALL ON)
+				endif()
```
Allows tests to be built when compiling for QNX with option QNX_BUILD_TESTS

dellaert replied:
Is there something specific to QNX that you need to use an environment variable? We normally do "make check" to run the tests.

JaiXJM-BB replied:
I think this can now be controlled by setting GTSAM_BUILD_TESTS to OFF, so no need for a QNX version. Removed env variable.

JaiXJM-BB commented on cmake/GtsamTesting.cmake:
```diff
@@ -240,8 +242,10 @@ macro(gtsamAddTestsGlob_impl groupName globPatterns excludedFiles linkLibraries)
-
+			if(NOT QNX OR NOT DEFINED ENV{QNX_BUILD_TESTS})
+				set_target_properties(${target_name} PROPERTIES EXCLUDE_FROM_ALL ON)
+			endif()
```
Allows tests to be built when compiling for QNX with option QNX_BUILD_TESTS

jlblancoc replied:
Why is QNX in need of special handling here for tests?

JaiXJM-BB commented on gtsam/3rdparty/ceres/CMakeLists.txt:
```diff
@@ -1,2 +1,7 @@
+else()
+    #Install in the install include directory rather than at the install prefix.
+    install(FILES ${ceres_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/gtsam/3rdparty/ceres)
+endif()
```
There's a number of changes identical to this one, which handle the correct include directory for installation in qnx. Note that CMAKE_INSTALL_INCLUDEDIR should point to include/ by default, so in the future the QNX line could just replace the default one.

dellaert replied:
What is `CMAKE_INSTALL_INCLUDEDIR` now by default? If incorrect, should we do a PR on that first?

JaiXJM-BB commented on gtsam/CMakeLists.txt:
```diff
@@ -99,6 +99,10 @@ IF(MSVC)
+#For config.in searches
+if(QNX)
+set(QNX_TARGET_DATASET_DIR "$ENV{QNX_TARGET_DATASET_DIR}")
+endif()
```
Dataset directory is not examples/Data when running tests out of the source tree, so this is required to properly set the path.

jlblancoc replied:
Can you please explain why this is needed only for qnx? If they are example/data files shipped with gtsam, we should use CMAKE_SOURCE_DIRECTORY or alike instead.

JaiXJM-BB commented on gtsam/base/CMakeLists.txt:
```diff
@@ -1,9 +1,17 @@
-
 file(GLOB base_headers_tree "treeTraversal/*.h")
-install(FILES ${base_headers_tree} DESTINATION include/gtsam/base/treeTraversal)
+
```
slightly rearranged for convenience, but same as the other include directory changes.

JaiXJM-BB commented on gtsam/base/std_optional_serialization.h:
```diff
@@ -52,6 +52,10 @@ namespace boost { namespace serialization { struct U; } }
+// QCC (The QNX GCC-based Compiler) also has this issue, but it also extends to trivial destructor.
+#if defined(__QNX__)
+namespace std { template<> struct is_trivially_destructible<boost::serialization::U> : std::false_type {}; }
+#endif
```
The https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84075 issue extendes to QNX, but also affects `is_trivially_destructible`. Luckily it can be fixed in the same way.

JaiXJM-BB commented on gtsam/config.h.in:
```diff
@@ -26,7 +26,12 @@
+//Set toolbox path to the path on the target.
+#define GTSAM_INSTALLED_DATASET_DIR "@QNX_TARGET_DATASET_DIR@/gtsam_examples/Data"
+#endif
 
```
Data path changed when testing outside of the source tree on QNX.

JaiXJM-BB commented on gtsam/discrete/CMakeLists.txt:
```diff
@@ -1,7 +1,12 @@
 set(subdir discrete)
 file(GLOB discrete_headers "*.h")
-install(FILES ${discrete_headers} DESTINATION include/gtsam/discrete)
+# FIXME: exclude headers
```
This FIXME was not added by me; so I assume it may have been changed in the 430 or so commits I needed to merge into this one to get up to develop. This should be double checked before merging.

JaiXJM-BB commented on gtsam/inference/tests/testOrdering.cpp:
```diff
@@ -303,7 +303,7 @@ TEST(Ordering, MetisLoop) {
 
   // METIS
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(__QNX__)
```
Follows apple ordering in test output.

JaiXJM-BB commented on gtsam/nonlinear/tests/testSerializationNonlinear.cpp:
```diff
@@ -153,9 +153,15 @@ TEST(Serialization, NoiseModelFactor1_backwards_compatibility) {
                          factor_deserialized_xml);
+  #else
+  bool c = deserializeFromXMLFile(
+                        "priorFactor.xml",
```
Again, a filepath search error on QNX. Changed to look in the same directory now, as outside of source tree.

JaiXJM-BB commented on gtsam/symbolic/tests/testSymbolicBayesTree.cpp:
```diff
@@ -687,7 +687,7 @@ TEST(SymbolicBayesTree, COLAMDvsMETIS) {
     Ordering ordering = Ordering::Create(Ordering::METIS, sfg);
 // Linux and Mac split differently when using Metis
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(__QNX__)
```
Again, ordering results on QNX seem to follow the APPLE results from Metis.

JaiXJM-BB commented on gtsam_unstable/discrete/tests/testScheduler.cpp:
```diff
@@ -140,7 +140,11 @@ TEST(schedulingExample, test) {
+  #if !defined(__QNX__)
   string path(TOPSRCDIR "/gtsam_unstable/discrete/examples/");
+  #else
+  string path(""); //Same Directory
```
another path change when testing out of the source tree.

dellaert replied:
Hmmm, maybe this should be done with findExampleDataFile?

JaiXJM-BB commented on gtsam_unstable/partition/tests/testFindSeparator.cpp:
```diff
@@ -104,7 +104,13 @@ TEST ( Partition, edgePartitionByMetis2 )
   std::vector<size_t> keys{0, 1, 2, 3, 4};
+  #else
+  //Anything where 2 is before 0 will work.
+  std::vector<size_t> keys{2, 0, 3, 1, 4};
```
This error was quite interesting; but the summary is that 0 1 2 3 4 will partition into 3-4, 0-1-2 instead of 0-1, 2-3-4 on QNX. This is an equally valid partition, it is just interesting that the tiebreaker seems to be related to ordering here.

dellaert replied:
Yeah - we had some similar things in the past :-)

dellaert commented on gtsam/3rdparty/ceres/CMakeLists.txt:
```diff
@@ -1,2 +1,7 @@
+else()
+    #Install in the install include directory rather than at the install prefix.
+    install(FILES ${ceres_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/gtsam/3rdparty/ceres)
+endif()
```
Add newline at end

dellaert commented on gtsam/3rdparty/metis/include/CMakeLists.txt:
```diff
@@ -1 +1,6 @@
+    install(FILES metis.h DESTINATION include/gtsam/3rdparty/metis)
+else()
+    #Install in the install include directory rather than at the install prefix.
+    install(FILES metis.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/gtsam/3rdparty/metis)
```
same comment. And, add newline at end :-)

dellaert reviewed (commented):
Two big comments, questions:

- Since there are so many changes related to CMAKE_INSTALL_INCLUDEDIR, could we not fix that first in a PR first, if it is fixable?
- Can you explain a bit more as to why QNX needs special handling of tests, and how?

dellaert commented:
Re-reading your PR comment: indeed, I'd rather not bring in environment variables as a way to configure, as I don't think we use them in any other case - except maybe if third party libraries look for them.

JaiXJM-BB commented:
> Re-reading your PR comment: indeed, I'd rather not bring in environment variables as a way to configure, as I don't think we use them in any other case - except maybe if third party libraries look for them.

Sounds good; there is nothing that requires an environment variable on QNX's end - these could easily be a CMake Flag or removed entirely.

With regards to `make check`, this doesn't work for QNX (or at least in my experience I've never seen it done or gotten it to work) since ctest runs the files on whatever machine is compiling, and there isn't a CMake port on QNX or a make utility available either, hence cross-compilation. This is why tests are being compiled into executables, so that users can copy them over and run them manually. It does add a lot of compile time which is why I had them on a toggle.

That being said - We could keep unit test cmake changes on the QNX fork or in a patchfile in the QNX documentation repo -  that way environment variables/extra cmake options don't affect your develop branch and any user that does want unit test control can find it elsewhere.

JaiXJM-BB commented:
For `CMAKE_INSTALL_INCLUDEDIR`, my understanding is that it defaults to `include` per https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html#module:GNUInstallDirs

Making a separate PR first to fix these is a good idea- there's also CMAKE_INSTALL_LIBDIR, CMAKE_INSTALL_BINDIR, and more that apply as well. Not sure if those are already in use.

JaiXJM-BB committed:
Merge branch 'borglab:develop' into develop

dellaert commented:
Gotcha. Take a look at the windows testing: there we created one executable per “folder”, which cut down on compile time quite a bit.

dellaert commented:
It seems the default is indeed `include`, so it seems that the code in your `else` clause is the correct way, right? In that case, maybe we don't need a new PR, just use remove the if-then-else and we'll merge this new correct way instead. @jlblancoc am I missing something?

JaiXJM-BB committed:
Updated QNX-relevant paths, removed environment variables. (#2)

Updated QNX-relevant paths (will check others next), removed environment variables.

JaiXJM-BB commented on cmake/GtsamBuildTypes.cmake:
```diff
@@ -207,7 +207,8 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang")
 
-if (NOT MSVC)
+if ((NOT MSVC) AND (NOT QNX))
+  option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" ON)
```
Think this might be a merging error - is line 211 meant to be here?

JaiXJM-BB commented on tests/testSerializationSlam.cpp:
```diff
@@ -634,7 +634,6 @@ TEST(SubgraphSolver, Solves) {
     std::map<Key, Vector> lambda;
     system.build(Ab, keyInfo, lambda);
-
     // Create a perturbed (non-zero) RHS
```
erroneous, will fix soon

JaiXJM-BB committed:
Fixed remaining DESTINATION include/*

JaiXJM-BB committed:
removed extra whitespace

JaiXJM-BB committed:
Fixing whitespace

dellaert commented:
I see you are updating. re-iterating, and seeking @jlblancoc approval, I'd just use CMAKE_INSTALL_INCLUDEDIR in this PR, not bothering with the if-then-else statements, or another PR.

JaiXJM-BB commented:
@dellaert Sorry for the delay; I've just pushed a bunch of changes per discussion above.

-> Header files with `install( abc DESTINATION include/xyz)` should now all be `install( abc DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/xyz)`

-> Fixed the missing trailing whitespaces and some additional internal ones

->Environment Variables removed:
---> Environment variable `QNX_BUILD_TESTS` was redundant to `GTSAM_BUILD_TESTS` so I removed it entirely
---> `QNX_TARGET_DATASET_DIR` changed to a cmake cache variable, it can be set with `-DQNX_TARGET_DATASET_DIR:STRING="filepath"` and will default to the free community licenses' home directory.

Edit: just saw your message, I think I made sure to do that so its simpler.

dellaert commented:
This looks good to me but @jlblancoc might see something I don't

jlblancoc reviewed (changes requested):
Good in general. A couple comments on the test directories, which I don't understand... may need a few more details on why does it fail? We all build gtsam out-of-source and have no issues on other OSes.

JaiXJM-BB commented:
> Good in general. A couple comments on the test directories, which I don't understand... may need a few more details on why does it fail? We all build gtsam out-of-source and have no issues on other OSes.

QNX Doesn't have a native compiler - or at least, hasn't had one officially supported in many years, and not on the free non-commercial/research license which is a newer version. This makes building in the source tree on target not possible for the average user. The official way is to compile using our cross compiler "qcc" (a gcc wrapper) on Linux or Windows, then transfer the libraries/executables over.

This is generally not an issue, but in this case the tests pose two problems for QNX: 
1. The tests all follow paths on the source tree, so running them on-target would require building on Linux/Windows then copying the entire source tree over via SSH or equivalent method. This is not ideal since QNX's largest use case is on embedded targets which can have limited storage constraints (ex. our default VM for testing can't even fit boost's compiled libraries)
2. GTSAM's tests are ran using `make check` via` ctest`. We don't have `CMake`, `CTest`, or `make` on QNX, so it isn't possible to run tests like this even if the source tree were on target. 

To make tests more viable for running on QNX, my solution is to compile the tests into executables and change filepaths to the same directory, so users can just copy over the tests and required data. I'm open to changing this solution, but it just seemed to be ideal from my experience.

jlblancoc commented:
@JaiXJM-BB Oh, now I understand. Then, LGTM. This is quite a special case :-) 
I would just suggest you adding a brief (1-2 lines?) summary of this motivation close to some of those cmake places where you introduced the special behavior for QNX, so future users browsing cmake would understand it. 

Once that is done, the rest lgtm too, and @dellaert  could merge! :+1:

JaiXJM-BB committed:
Added justification for CMake changes, with reference to this PR.

JaiXJM-BB commented:
> @JaiXJM-BB Oh, now I understand. Then, LGTM. This is quite a special case :-) I would just suggest you adding a brief (1-2 lines?) summary of this motivation close to some of those cmake places where you introduced the special behavior for QNX, so future users browsing cmake would understand it.
> 
> Once that is done, the rest lgtm too, and @dellaert could merge! 👍

Should be good with 4367d94 !

dellaert commented:
Merged. welcome QNX!!!
